File: app/src/main/java/ru/stwtforever/fast/MainActivity.java
package ru.stwtforever.fast;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.view.View;
import android.widget.AdapterView;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.SlidingDrawer;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;
import java.util.List;

import ru.stwtforever.fast.adapter.DrawerAdapter;
import ru.stwtforever.fast.api.UserConfig;
import ru.stwtforever.fast.api.VKApi;
import ru.stwtforever.fast.api.model.VKUser;
import ru.stwtforever.fast.cls.BaseFragment;
import ru.stwtforever.fast.common.ThemeManager;
import ru.stwtforever.fast.concurrent.AsyncCallback;
import ru.stwtforever.fast.concurrent.ThreadExecutor;
import ru.stwtforever.fast.db.DBHelper;
import ru.stwtforever.fast.db.MemoryCache;
import ru.stwtforever.fast.fragment.FragmentDialogs;
import ru.stwtforever.fast.fragment.FragmentFriends;
import ru.stwtforever.fast.fragment.FragmentGroups;
import ru.stwtforever.fast.helper.DialogHelper;
import ru.stwtforever.fast.helper.PermissionHelper;
import ru.stwtforever.fast.service.LongPollService;
import ru.stwtforever.fast.util.ArrayUtil;
import ru.stwtforever.fast.util.Utils;
import ru.stwtforever.fast.util.ViewUtils;

public class MainActivity extends AppCompatActivity {

    private View bottom_toolbar;

    private LinearLayout tb_btn_switch;
    private ImageButton messages, friends, filter, menu;

    private GradientDrawable bg;

    private SlidingDrawer drawer;
    private DrawerAdapter adapter;

    private FragmentDialogs f_dialogs = new FragmentDialogs();
    private FragmentFriends f_friends = new FragmentFriends();
    private FragmentGroups f_groups = new FragmentGroups();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        PermissionHelper.init(this);
        ViewUtils.applyWindowStyles(this);
        EventBus.getDefault().register(this);
        setTheme(ThemeManager.getCurrentTheme());

        VKApi.config = UserConfig.restore();

        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        bottom_toolbar = findViewById(R.id.toolbar);

        drawer = findViewById(R.id.drawer);
        initToolbar();

        drawer.setOnDrawerCloseListener(new SlidingDrawer.OnDrawerCloseListener() {

            @Override
            public void onDrawerClosed() {
                drawer.setVisibility(View.GONE);
            }

        });

        drawer.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View p1) {
                drawer.animateClose();
            }
        });

        checkLogin();

        startService(new Intent(this, LongPollService.class));
        checkCrash();

        if (UserConfig.isLoggedIn()) {
            trackVisitor();
        }
    }

    private void startLoginActivity() {
        startActivity(new Intent(this, LoginActivity.class));
        finish();
    }

    private void checkLogin() {
        UserConfig.restore();
        if (!UserConfig.isLoggedIn()) {
            startLoginActivity();
        } else {
            replaceFragment(f_dialogs);
        }
    }

    private void trackVisitor() {
        ThreadExecutor.execute(new AsyncCallback(this) {

            int i;

            @Override
            public void ready() throws Exception {
                i = VKApi.stats().trackVisitor().execute(Integer.class).get(0);
            }

            @Override
            public void done() {
            }

            @Override
            public void error(Exception e) {
                e.printStackTrace();
            }

        });
    }

    private void checkCrash() {
        if (Utils.getPrefs().getBoolean("isCrashed", false)) {
            final String trace = Utils.getPrefs().getString("crashLog", "");
            Utils.getPrefs().edit().putBoolean("isCrashed", false).putString("crashLog", "").apply();


            AlertDialog.Builder adb = new AlertDialog.Builder(this);
            adb.setTitle(R.string.warning);

            String message = getString(R.string.app_crashed);

            adb.setMessage(message);
            adb.setPositiveButton(android.R.string.ok, null);
            adb.setNeutralButton(R.string.show_error, new DialogInterface.OnClickListener() {

                @Override
                public void onClick(DialogInterface p1, int p2) {
                    AlertDialog.Builder adb = new AlertDialog.Builder(MainActivity.this);
                    adb.setTitle(R.string.error_log);
                    adb.setMessage(trace);
                    adb.setPositiveButton(android.R.string.ok, null);
                    adb.setNeutralButton(R.string.copy, new DialogInterface.OnClickListener() {

                        @Override
                        public void onClick(DialogInterface p1, int p2) {
                            Utils.copyText(trace);
                        }
                    });
                    DialogHelper.create(adb).show();
                }

            });
            AlertDialog alert = DialogHelper.create(adb);
            alert.show();
        }
    }

    private void initToolbar() {
        filter = bottom_toolbar.findViewById(R.id.tb_filter);
        menu = bottom_toolbar.findViewById(R.id.tb_menu);

        filter.setEnabled(false);
        filter.getDrawable().setTint(Color.GRAY);

        menu.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                drawer.setVisibility(View.VISIBLE);
                drawer.animateToggle();
            }

        });

        bg = new GradientDrawable();
        bg.setColor(ThemeManager.getAccent());
        bg.setCornerRadius(200);

        messages = bottom_toolbar.findViewById(R.id.tb_messages);
        friends = bottom_toolbar.findViewById(R.id.tb_friends);

        messages.setBackground(bg);
        friends.setBackgroundColor(Color.TRANSPARENT);

        friends.setOnClickListener(click);
        messages.setOnClickListener(click);

        tb_btn_switch = bottom_toolbar.findViewById(R.id.tb_icons_switcher);
        tb_btn_switch.setBackgroundResource(ThemeManager.isDark() ? R.drawable.tb_switcher_bg_dark : R.drawable.tb_switcher_bg);

        ListView list = findViewById(R.id.drawer_list);

        final String[] titles = new String[]{getString(R.string.groups), getString(R.string.settings), getString(R.string.logout)};
        int[] icons = new int[]{R.drawable.ic_people_outline_black_24dp, R.drawable.md_settings, R.drawable.md_exit_to_app};
        final String[] tags = new String[]{"groups", "settings", "exit"};

        ArrayList<Object[]> items = new ArrayList<>();

        for (int i = 0; i < titles.length; i++) {
            items.add(new Object[]{titles[i], icons[i], false, tags[i]});
        }

        adapter = new DrawerAdapter(this, items);
        list.setAdapter(adapter);

        list.setOnItemClickListener(new ListView.OnItemClickListener() {

            @Override
            public void onItemClick(AdapterView<?> lv, View v, int position, long p4) {
                boolean exitOrSettings = (tags[position].equals("settings") || tags[position].equals("exit"));

                if (!exitOrSettings)
                    adapter.setHover(true, position);

                itemClick(position);

                if (!tags[position].equals("settings")) {
                    drawer.animateClose();
                } else {
                    drawer.close();
                }
            }
        });
    }

    private void itemClick(int position) {
        String tag = (String) adapter.getValues().get(position)[3];

        switch (tag) {
            case "groups":
                //replaceFragment(f_groups);
                break;
            case "settings":
                startActivity(new Intent(this, SettingsActivity.class));
                break;
            case "exit":
                showExitDialog();
                break;
        }
    }

    private View.OnClickListener click = new View.OnClickListener() {

        @Override
        public void onClick(View v) {
            if (adapter == null) return;
            adapter.clearHover();

            Fragment f = null;

            switch (v.getId()) {
                case R.id.tb_messages:
                    f = f_dialogs;
                    messages.setBackground(bg);
                    messages.getDrawable().setTint(Color.WHITE);
                    friends.getDrawable().setTint(ThemeManager.getAccent());
                    friends.setBackgroundColor(Color.TRANSPARENT);

                    replaceFragment(f_dialogs);
                    break;
                case R.id.tb_friends:
                    f = f_friends;
                    friends.getDrawable().setTint(Color.WHITE);
                    messages.getDrawable().setTint(ThemeManager.getAccent());
                    friends.setBackground(bg);
                    messages.setBackgroundColor(Color.TRANSPARENT);

                    replaceFragment(f_friends);
                    break;
            }

            if (getVisibleFragment() == f) {
                ((BaseFragment) f).getList().scrollToPosition(0);
            }
        }
    };

    @Subscribe(sticky = true)
    public void onUpdateUser(VKUser u) {
        return;
    }

    private void showExitDialog() {
        AlertDialog.Builder adb = new AlertDialog.Builder(this);
        adb.setTitle(R.string.warning);
        adb.setMessage(R.string.exit_message);
        adb.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                UserConfig.clear();
                startLoginActivity();
            }
        });
        adb.setNegativeButton(R.string.no, null);
        AlertDialog alert = adb.create();
        alert.show();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
        MemoryCache.clear();
        DBHelper.getInstance().close();
    }

    private void replaceFragment(Fragment fragment) {
        FragmentTransaction tr = getSupportFragmentManager().beginTransaction();

        List<Fragment> fragments = getSupportFragmentManager().getFragments();

        if (ArrayUtil.isEmpty(fragments)) {
            tr.add(R.id.fragment_container, fragment, fragment.getClass().getSimpleName());
        } else {
            boolean exists = false;

            for (Fragment f : fragments) {
                if (f.getClass().getSimpleName().equals(fragment.getClass().getSimpleName())) {
                    tr.show(f);
                    exists = true;
                } else {
                    tr.hide(f);
                }
            }

            if (!exists) {
                tr.add(R.id.fragment_container, fragment, fragment.getClass().getSimpleName());
            }
        }

        tr.commit();
    }

    private Fragment getVisibleFragment() {
        List<Fragment> frs = getSupportFragmentManager().getFragments();

        if (ArrayUtil.isEmpty(frs)) return null;

        Fragment f = null;

        for (Fragment fr : frs) {
            if (fr.isVisible()) {
                f = fr;
            }
        }

        return f;
    }
}

File: app/src/main/java/ru/stwtforever/fast/MainActivity.kt
package ru.stwtforever.fast

import android.Manifest
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.ImageButton
import android.widget.LinearLayout
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import ru.stwtforever.fast.api.UserConfig
import ru.stwtforever.fast.api.VKApi
import ru.stwtforever.fast.api.model.VKUser
import ru.stwtforever.fast.cls.BaseFragment
import ru.stwtforever.fast.common.ThemeManager
import ru.stwtforever.fast.concurrent.AsyncCallback
import ru.stwtforever.fast.concurrent.ThreadExecutor
import ru.stwtforever.fast.database.DatabaseHelper
import ru.stwtforever.fast.database.MemoryCache
import ru.stwtforever.fast.fragment.FragmentDialogs
import ru.stwtforever.fast.fragment.FragmentFriends
import ru.stwtforever.fast.fragment.FragmentGroups
import ru.stwtforever.fast.fragment.FragmentNavDrawer
import ru.stwtforever.fast.helper.DialogHelper
import ru.stwtforever.fast.helper.PermissionHelper
import ru.stwtforever.fast.service.LongPollService
import ru.stwtforever.fast.util.ArrayUtil
import ru.stwtforever.fast.util.Utils
import ru.stwtforever.fast.util.ViewUtils

class MainActivity : AppCompatActivity() {

    private var bottom_toolbar: View? = null

    private var tb_btn_switch: LinearLayout? = null
    private var messages: ImageButton? = null
    private var friends: ImageButton? = null
    private var filter: ImageButton? = null
    private var menu: ImageButton? = null

    private var bg: GradientDrawable? = null

    private val f_dialogs = FragmentDialogs()
    private val f_friends = FragmentFriends()
    private val f_groups = FragmentGroups()

    private var selected_id = -1

    private val click = View.OnClickListener { v ->
        selected_id = -1
        var f: Fragment? = null

        when (v.id) {
            R.id.tb_messages -> {
                f = f_dialogs
                messages!!.background = bg
                messages!!.drawable.setTint(Color.WHITE)
                friends!!.drawable.setTint(ThemeManager.getAccent())
                friends!!.setBackgroundColor(Color.TRANSPARENT)

                replaceFragment(f_dialogs)
            }
            R.id.tb_friends -> {
                f = f_friends
                friends!!.drawable.setTint(Color.WHITE)
                messages!!.drawable.setTint(ThemeManager.getAccent())
                friends!!.background = bg
                messages!!.setBackgroundColor(Color.TRANSPARENT)

                replaceFragment(f_friends)
            }
        }

        if (visibleFragment === f) {
            (f as BaseFragment).recyclerView.scrollToPosition(0)
        }
    }

    private val visibleFragment: Fragment?
        get() {
            val frs = supportFragmentManager.fragments

            if (ArrayUtil.isEmpty(frs)) return null

            var f: Fragment? = null

            for (fr in frs) {
                if (fr.isVisible) {
                    f = fr
                }
            }

            return f
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        PermissionHelper.init(this)
        ViewUtils.applyWindowStyles(this)
        EventBus.getDefault().register(this)
        setTheme(ThemeManager.getCurrentTheme())

        VKApi.config = UserConfig.restore()

        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        bottom_toolbar = findViewById(R.id.toolbar)

        initToolbar()

        checkLogin()

        startService(Intent(this, LongPollService::class.java))
        checkCrash()

        if (UserConfig.isLoggedIn()) {
            trackVisitor()
        }

        if (!PermissionHelper.isGrantedPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
            PermissionHelper.requestPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, 44)
        }
    }

    private fun startLoginActivity() {
        startActivity(Intent(this, LoginActivity::class.java))
        finish()
    }

    private fun checkLogin() {
        UserConfig.restore()
        if (!UserConfig.isLoggedIn()) {
            startLoginActivity()
        } else {
            replaceFragment(f_dialogs)
        }
    }

    private fun trackVisitor() {
        ThreadExecutor.execute(object : AsyncCallback(this) {

            internal var i: Int = 0

            @Throws(Exception::class)
            override fun ready() {
                i = VKApi.stats().trackVisitor().execute<Int>(Int::class.java)[0]
            }

            override fun done() {}

            override fun error(e: Exception) {
                e.printStackTrace()
            }

        })
    }

    private fun checkCrash() {
        if (Utils.getPrefs().getBoolean("isCrashed", false)) {
            val trace = Utils.getPrefs().getString("crashLog", "")
            Utils.getPrefs().edit().putBoolean("isCrashed", false).putString("crashLog", "").apply()


            val adb = AlertDialog.Builder(this)
            adb.setTitle(R.string.warning)

            adb.setMessage(R.string.app_crashed)
            adb.setPositiveButton(android.R.string.ok, null)
            adb.setNeutralButton(R.string.show_error) { _, _ ->
                val adb = AlertDialog.Builder(this@MainActivity)
                adb.setTitle(R.string.error_log)
                adb.setMessage(trace)
                adb.setPositiveButton(android.R.string.ok, null)
                adb.setNeutralButton(R.string.copy) { _, _ -> Utils.copyText(trace) }
                DialogHelper.create(adb).show()
            }
            val alert = DialogHelper.create(adb)
            alert.show()
        }
    }

    private fun initToolbar() {
        filter = bottom_toolbar!!.findViewById(R.id.tb_filter)
        menu = bottom_toolbar!!.findViewById(R.id.tb_menu)

        filter!!.isEnabled = false
        filter!!.drawable.setTint(Color.GRAY)

        menu!!.setOnClickListener {
            val drawer = FragmentNavDrawer()

            val b = Bundle()
            b.putInt("selected_id", selected_id)

            drawer.arguments = b
            drawer.show(supportFragmentManager, "")
        }

        bg = GradientDrawable()
        bg!!.setColor(ThemeManager.getAccent())
        bg!!.cornerRadius = 200f

        messages = bottom_toolbar!!.findViewById(R.id.tb_messages)
        friends = bottom_toolbar!!.findViewById(R.id.tb_friends)

        messages!!.background = bg
        friends!!.setBackgroundColor(Color.TRANSPARENT)

        friends!!.setOnClickListener(click)
        messages!!.setOnClickListener(click)

        tb_btn_switch = bottom_toolbar!!.findViewById(R.id.tb_icons_switcher)
        tb_btn_switch!!.setBackgroundResource(if (ThemeManager.isDark()) R.drawable.tb_switcher_bg_dark else R.drawable.tb_switcher_bg)
    }

    fun itemClick(item: MenuItem) {
        when (item.itemId) {
            R.id.groups -> selected_id = R.id.groups
            R.id.settings -> startActivity(Intent(this, SettingsActivity::class.java))
            R.id.left -> showExitDialog()
        }
    }

    @Subscribe(sticky = true)
    fun onUpdateUser(u: VKUser) {
        return
    }

    private fun showExitDialog() {
        val adb = AlertDialog.Builder(this)
        adb.setTitle(R.string.warning)
        adb.setMessage(R.string.exit_message)
        adb.setPositiveButton(R.string.yes) { dialogInterface, i ->
            UserConfig.clear()
            startLoginActivity()
        }
        adb.setNegativeButton(R.string.no, null)
        val alert = adb.create()
        alert.show()
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
        MemoryCache.clear()
        DatabaseHelper.getInstance().close()
    }

    private fun replaceFragment(fragment: Fragment) {
        val tr = supportFragmentManager.beginTransaction()

        val fragments = supportFragmentManager.fragments

        if (ArrayUtil.isEmpty(fragments)) {
            tr.add(R.id.fragment_container, fragment, fragment.javaClass.getSimpleName())
        } else {
            var exists = false

            for (f in fragments) {
                if (f.javaClass.getSimpleName() == fragment.javaClass.getSimpleName()) {
                    tr.show(f)
                    exists = true
                } else {
                    tr.hide(f)
                }
            }

            if (!exists) {
                tr.add(R.id.fragment_container, fragment, fragment.javaClass.getSimpleName())
            }
        }

        tr.commit()
    }
}
